src/backend/opencl/solve.cpp

```diff
@@ -160,7 +160,12 @@ Array<T> leastSquares(const Array<T> &a, const Array<T> &b) {
                            A.getOffset(), A.strides()[1], &h_tau[0], (*dT)(),
                            tmp.getOffset(), NB, queue, &info);
 
-        B = matmul(A, B, AF_MAT_NONE, AF_MAT_NONE);
+        int B_new_dim2 = A.dims()[2] >= B.dims()[2] ? A.dims()[2] : B.dims()[2];
+        int B_new_dim3 = A.dims()[3] >= B.dims()[3] ? A.dims()[3] : B.dims()[3];
+        Array<T> B_new = createEmptyArray<T>(dim4(A.dims()[0], B.dims()[1],
+                                                  B_new_dim2, B_new_dim3));
+        matmul(B_new, A, B, AF_MAT_NONE, AF_MAT_NONE);
```

Comment:
Using the latest change, you can even avoid calculating output dimensions here and let matmul do the allocation.